spec/fastsync/version2/fsv2_scheduler.tla

```diff
@@ -0,0 +1,532 @@
+Init == turn = "environment" /\ InitEnv /\ InitSc
+
+FlipTurn == 
+  turn' = (
+    IF turn = "scheduler" THEN 
+      "environment" 
+    ELSE
+      "scheduler"
+  ) 
+
+Next == \* scheduler and environment alternate their steps (synchronous composition introduces more states)
+  /\ FlipTurn
+  /\ IF turn = "scheduler" THEN 
+       /\ NextSc
+       /\ inEvent' = noEvent
+       /\ UNCHANGED envRunning
+     ELSE
+       /\ NextEnv
+       /\ outEvent' = noEvent
+       /\ UNCHANGED <<scRunning, scheduler>>  
+
+(* ----------------------------------------------------------------------------------------------*)
+(* Expected properties                                                                           *)
+(* ----------------------------------------------------------------------------------------------*)
+
+TypeOK ==
+    /\ turn \in {"scheduler", "environment"}
+    /\ inEvent \in InEvents
+    /\ envRunning \in BOOLEAN
+    /\ outEvent \in OutEvents
+    /\ scheduler \in [
+         initHeight: Heights \cup {ultimateHeight + 1},
+         height: Heights \cup {ultimateHeight + 1},
+         peers: SUBSET PeerIDs,
+         peerHeights: [PeerIDs -> Heights \cup {None}],
+         peerStates: [PeerIDs -> PeerStates],
+         blocks: SUBSET Heights,
+         blockStates: [Heights -> BlockStates],
+         pendingBlocks: [Heights -> PeerIDs \cup {None}],
+         receivedBlocks: [Heights -> PeerIDs \cup {None}]
+       ]
+       
+TerminationWhenNoAdvance ==
+  (WF_turn(FlipTurn) /\ <> (inEvent.type = "tNoAdvanceExp"))
+    => <> ~scRunning
+
+NoFailuresAndTimeouts ==
+    \* no peer removal
+    /\ inEvent.type /= "peerRemoveEv"
+    \* no invalid blocks
+    /\ inEvent.type /= "bcNoBlockResponse"
+    /\ inEvent.type /= "pcBlockVerificationFailure"
+NoGlobalTimeout == inEvent.type /= "tNoAdvanceExp"
+
+FiniteResponse ==
+    <>[] (inEvent.type
+         \in {"bcAddNewPeer", "bcStatusResponse", "bcBlockResponse", "pcBlockProcessed", "pcBlockVerificationFailure", "rTrySchedule"})
+
+\* all blocks from initHeight up to max peer height have been processed
+AllRequiredBlocksProcessed ==
+  LET maxH == maxHeightScheduler(scheduler) IN
+  LET unprocessedBlocks == {\A h \in scheduler.initHeight.. maxH: scheduler.blockStates[h] /= "blockStateProcessed"} IN
+  scheduler.height >= maxH /\ Cardinality(unprocessedBlocks) = 0
+
+GoodTermination ==
+  (WF_turn(FlipTurn) /\ []NoGlobalTimeout /\ []NoFailuresAndTimeouts /\ FiniteResponse)
+     => <>(outEvent.type = "ScFinishedEv" /\ AllRequiredBlocksProcessed)
+
+\* TougherTermination is the termination property we like more and it holds true.
+\* When IncreaseHeight holds true, fastsync is progressing, unless it has hit the ceiling.
+IncreaseHeight ==
+    scheduler'.height > scheduler.height \/ scheduler'.height >= ultimateHeight
+
+TougherTermination ==
+  (scheduler.height < ultimateHeight
+    /\ (WF_turn(FlipTurn)
+    /\ SF_inEvent(OnTrySchedule)
+    /\ FiniteResponse
```

Comment:
So the idea here is there could be some failures, but since there's a fixed number of peers eventually we'll detect all the failures, get rid of those peers, and be left with only good peers and hence no more failures. And then we'll be able to terminate.

Is that right? But then why should FiniteResponse include pcBlockVerificationFailure?

Reply:
> So the idea here is there could be some failures, but since there's a fixed number of peers eventually we'll detect all the failures, get rid of those peers, and be left with only good peers and hence no more failures. And then we'll be able to terminate.
> 
> Is that right? 

Yes.

>But then why should FiniteResponse include pcBlockVerificationFailure?

I was trying with 2 peers and leave this in to show that we never terminate if both peers are removed (model, like Go code, doesn't allow to add them back). Not able to show this :( still trying to figure it out.

Reply:
Ah ok!